rfcs/web_extensions.md

```diff
@@ -0,0 +1,199 @@
+Asserts that a given condition is `false`. If the condition is `true`, the test fails.
+
+**Parameters**
+
+* `condition` (boolean) – The condition to assert. If `true`, the test fails.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertEq(actual, expected, message)`**\
+Throws an assertion error and fails the test if the actual value is not _logically_ equal to the
+expected value, by these rules:
+
+1) _Plain arrays_ (`Array.isArray()` returns true) are compared for _deep_ equality.
+    1.1) Their `length` and all items are recursively compared by these same rules.
+2) _Plain objects_ (with prototype `Object.prototype` or `null`) are also compared recursively.
+    2.1) Only enumerable _own_ properties are considered, unordered.
+3) Everything else is compared using `Object.is()` [same value algorithm](https://tc39.es/ecma262/#sec-samevalue).
+
+We're starting with support for the minimal set of compound types, and plan to expand custom rules
+to include `ArrayBuffer`s, `Map`s, `Set`s, etc. in congruence with Node's [assert.deepStrictEqual()](https://nodejs.org/api/assert.html#comparison-details_1).
+
+**Parameters**
+
+* `actual` (any) – The actual value.
+* `expected` (any) – The expected value.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertThrows(fn, expectedError, message)`**
+Asserts that a given function throws an error when executed. If the function does not throw, the test fails.
+The test will also fail if `expectedError` is defined and the function does not throw an error with a message matching `expectedError`.
+
+When `assertThrows` is invoked with two parameters, the second value should be interpreted as `expectedError`. This is explicitly mentioned because `expectedError` and `message` are both optional and accepting string values.
+
+**Parameters**
+
+* `fn` (function) – A function expected to throw an error.
+* `expectedError` (any, optional) – A string (which must match exactly) or a RegExp stating the expected error.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.runTests(tests)`**
+Queues an array of test functions to be executed sequentially and returns a promise that resolves or
+rejects based on the test outcomes.
+
+Each function in the `tests` array is treated as an individual test case. Tests are executed in the
+order they appear in the array. If multiple tests are passed, a new test begins once the previous
+tests completes. If a test fails—either due to a failed assertion or an unexpected exception—execution
+of that test stops, and the next test in the queue begins.
+
+After the returned promise is settled, this method can be called again to run additional tests.
+Calling it while test are still running throws or returns a rejected promise.
+
+The returned promise:
+- **Resolves** if all tests complete successfully.
+- **Rejects** if any test fails.
```

Comment:
This is obviously your existing API, but for web-platform-tests you really want to structure things so that tests are as independent as possible i.e. later tests running doesn't depend on earlier tests passing. We generally have the semantics that asserts fail the test and stop further execution of that one test, but as far as possible don't prevent future tests running. Otherwise you can get into a situation where a browser that would pass all but one of the subtests in a specific file actually stops after the first (failing) subtest.

Reply:
This documentation was moved [here ](https://github.com/w3c/webextensions/blob/main/proposals/browser_test_api.md) but the `runTests` method runs all of the tests, regardless if a previous test failed. The promise returned will reject if any test fails, but each test is guaranteed to run and the results for each test will be reported.

Reply:
Yes, that's our intention as well, we clarified the wording in https://github.com/w3c/webextensions/blob/main/proposals/browser_test_api.md